packages/taraxa-sdk/src/DposClient.ts

```diff
@@ -0,0 +1,296 @@
+import { ethers } from 'ethers';
+import { DposAbi, DposAbi__factory as Dpos } from '../types/ethers-contracts';
+import {
+  DPOS_CONTRACT_ADDRESS,
+  Network,
+  NetworkName,
+  ProviderType,
+  getNetwork,
+  getNetworkById,
+} from './networks';
+
+export class DposClient {
+  private _dpos: DposAbi;
+
+  private _provider: ethers.providers.Provider;
+
+  private _signer: ethers.Signer | null = null;
+
+  private _chainId: number | undefined;
+
+  private _network: Network | null;
+
+  constructor(
+    networkIdOrName: number | NetworkName,
+    providerType: ProviderType,
+    privateKey?: string
+  ) {
+    this._network = this.lookupNetwork(networkIdOrName);
+
+    if (!this._network) {
+      throw new Error('Invalid network ID or name');
+    }
+
+    this._chainId = this._network?.chainId;
+
+    if (!this._chainId) {
+      throw new Error('Chain ID not found for the provided network name');
+    }
+
+    this._provider = this.setupProvider(providerType, privateKey);
+    this._dpos = Dpos.connect(DPOS_CONTRACT_ADDRESS, this.provider);
+  }
+
+  public get dpos(): DposAbi {
+    return this._dpos;
+  }
+
+  public get signer(): ethers.Signer | null {
+    return this._signer;
+  }
+
+  public get provider(): ethers.providers.Provider {
+    return this._provider;
+  }
+
+  public get chain(): number | undefined {
+    return this._chainId;
+  }
+
+  public get network(): Network | null {
+    return this._network;
+  }
+
+  // Private methods
+  private setupProvider(
+    providerType: ProviderType,
+    privateKey?: string
+  ): ethers.providers.Provider {
+    let provider: ethers.providers.Provider;
+
+    if (providerType === ProviderType.RPC || typeof window === 'undefined') {
+      provider = new ethers.providers.JsonRpcProvider(this.network?.rpcUrl);
+      if (privateKey) {
+        this._signer = new ethers.Wallet(privateKey, provider);
+      }
+    } else if (
+      providerType === ProviderType.WEB3 &&
+      typeof window !== 'undefined' &&
+      window.ethereum
+    ) {
+      provider = new ethers.providers.Web3Provider(window.ethereum, 'any');
+      this._signer = (provider as ethers.providers.Web3Provider).getSigner();
+    } else {
+      throw new Error('Invalid provider type or environment');
+    }

```

Comment:
This if else chain feels a bit too verbose, we could use a nice little switch case here(I switched to this in my personal projects cuz its soo easy to read compared to)

```suggestion
    const conditionKey = `${providerType}_${typeof window}`;

switch (conditionKey) {
  case `${ProviderType.RPC}_undefined`:
    provider = new ethers.providers.JsonRpcProvider(this.network?.rpcUrl);
    if (privateKey) {
      this._signer = new ethers.Wallet(privateKey, provider);
    }
    break;

  case `${ProviderType.WEB3}_object`:
    if (window.ethereum) {
      provider = new ethers.providers.Web3Provider(window.ethereum, 'any');
      this._signer = provider.getSigner();
    }
    break;

  default:
    throw new Error('Invalid provider type or environment');
}

```

Reply:
It`s a lot more readable this way but it won`t handle the case where we are in an UI (React) app where we need an RPC connection but we also have window.ethereum.